rfcs/converged-composition-pattern.md

```diff
@@ -0,0 +1,168 @@
+| Accordion |   ✅   |         ✅          |      ✅      |    ✅     |
+| Menu      |   ✅   |         ✅          |      ❌      |    ✅     |
+| Button    |   ✅   |         ❌          |      ✅      |    ✅     |
+| Popover   |   ✅   |         ❌          |      ❌      |    ✅     |
+
+We do see all components have `state` and `render`. However, `styles` and/or `context` are sometimes added or omitted. Knowing which components use which pattern is not clear or documented.
+
+### 2. Complex imports
+
+Each `use*` function must be imported separately and assembled in the proper order. This makes a confusing user experience trying to recompose your own components.
+
+```tsx
+import { useFoo, useFooStyles, useFooContextValues, renderFoo } from '@fluentui/react-components';
+```
+
+Since the pattern is consistent, there is no way for the consumer to know what they need to import to create a component.
+They will need to consult the docs for each component.
+
+### 3. `useFoo` is not usable on its own
+
+The name implies I am using the full component, yet, without styles it is not truly unusable.
+
+It is also not intuitive as to why the `state` calculated by `useFoo` includes all props, but the `className` prop is handled separately in `useFooStyles`. Especially since there are other calls inside `useFoo` for resolving shorthand, calculating aria, and adding event listeners.
+
+### 4. Not learn-once
+
+Ideally, the overall story for our composition would be learn-once-apply-everywhere. Currently, many implementations have just a small difference from others.
+
+## Proposed Solution
+
+Prioritizing a learn-once approach and simplification, we can consider this API:
+
+```tsx
+const [state, render] = useComponent();
+
+// modify state
+
+return render(state);
+```
+
+Problems solved:
+
+1. **Inconsistent** - All components will have the same API since all components have `state` and `render`.
+2. **Complex Imports** - There is one import per component, named `useComponent`, opposed to 3-4 imports.
+3. **useFoo not usable** - "Using" a component returns a fully usable yet customizable component API.
+4. **Not learn-once** - Since this is consistent, and abstracts over all components, it is learn-once-apply-everywhere.
+
+### How is `useComponent` implemented?
+
+```tsx
+export const Component: ForwardRefComponent<ButtonProps> = React.forwardRef((props, ref) => {
+  const { state, render } = useButton(props, ref);
+
+  return render(state);
+}) as ForwardRefComponent<ButtonProps>;
+```
+
+The `useComponent` function would call all other state functions and return the state and the renderer:
+
+`useComponent`
+
+- `useComponentState`
+- `useComponentStyle`
+- `useComponentARIA`
+- return `[state, renderCompoonent]`
+
+## FAQ
+
+### How to update `state` before styles are calculated?
+
+If there is a use case where the consumer needs to get in between the state hooks, they can drop down a level and compose each hook together individually.
```

Comment:
I think updating state before styling will be a fairly common scenario, and it might be worth incorporating it into this solution. Otherwise, we're back with all of the problems listed in the Problem section.

I haven't figured out the best API for this yet, but I wonder if there's some way to model the set of hooks used by as a pipeline, where each hook modifies state and hands it to the next hook. The `useComponent` function would return the pipeline, and you could insert, remove, or modify the hooks that are called in the pipeline. We could have defined "sections" of the pipeline, like "applyStyle" or "render" that the dev could add hooks to run before them.

Again, the API _really_ (really) would need to be better thought out than this, but for example:
```ts
const buttonPipeline = useButton(props, ref);
buttonPipeline.insertBefore('applyStyle', (state) => {
  state.pressed = true;
});
buttonPipeline.insertBefore('render', (state) => {
  state.root.className += " foo";
});
return buttonPipeline.render();
```

The main benefit is that the user wouldn't have to explicitly know about each hook that's being called by `useButton`, but they have a way to add their own code in pre-defined places.

Reply:
> I haven't figured out the best API for this yet, but I wonder if there's some way to model the set of hooks used by as a pipeline, where each hook modifies state and hands it to the next hook. The `useComponent` function would return the pipeline, and you could insert, remove, or modify the hooks that are called in the pipeline. We could have defined "sections" of the pipeline, like "applyStyle" or "render" that the dev could add hooks to run before them.
> 

The meeting on this topic from earlier this week proposes an API that looks like:

```ts
const [state, render] = useComponent(props, 
// Optional list of overrides for hooks
{
    styles: (useStyles) => { /*...*/ }, // each function passes in the default hook. It probably also needs to pass in `state`
    ARIA: (useARIA) => { /* ... */ },
    // etc.
});
```

I'm still thinking about this proposal but what I like about it is that it gives users a nice way to hook into the component without us having to define a preset pipeline which can be nice because we may not want to always call the internal hooks (`useStyles`, `useARIA`, etc) in the same order. With this approach you can easily hook in before and after each hook with something like:

```ts
const [state, render] = useComponent(props, {
    styles: (useStyles) => {
        useMyBeforeStyles();
        useStyles(); // the default styling hook
        useMyAfterStyles();
    },
    // etc
});
```

The one part that is missing is how to handle inserting state modification before all internal hook calls? You could do this with my example above but that would require you knowing which hook is called first and if we're not defining an order that then requires users to have insider knowledge of the component which is not ideal. We could define a special hook for this but it will probably be noop most of the time. Maybe this isn't an issue because we always need the internal hooks to generate the default state for a component and we don't want to allow users to hook into this part of the pipeline?